### ISSUE TITLE
Discussion channel for how to apply self-play to custom env?

### ISSUE THREAD

**Issue by zxzzz0** [opendilab/DI-engine#5]
Hi all,

Nice project. We want to start using it. After reading the doc and the config `dizoo/competitive_rl/entry/cpong_dqn_default_config.py` for league train, there are still something not clear to us. Do you have a channel that can discuss trivial questions frequently? Like a WeChat group or slack channel? 

cc: opendilab.contact@gmail.com

**Comment by zxzzz0** [comment]
@PaParaZz1

**Comment by liuyuisanai** [comment]
@zxzzz0 You can ask our WeChat assistant for the group entry. Assistant WeChat ID: OpenDILab

**Comment by PaParaZz1** [comment]
> Hi all,
> 
> Nice project. We want to start use it. After reading the doc and the config `dizoo/competitive_rl/entry/cpong_dqn_default_config.py` for league train, there are still something not clear to us. Do you have a channel that can discuss trivial questions frequently? Like a WeChat group or slack channel?
> 
> cc: [opendilab.contact@gmail.com](mailto:opendilab.contact@gmail.com)

Thank you for your participation. And you can add our WeChat assistant ID listed above. Our slack channel will also be available soon. To be specific, what kinds of training tasks you want to do? I guess you want to train a self-play league in cpong environment, and `dizoo/competitive_rl/entry/cpong_dqn_default_config.py` is just a agent VS bot training config. If you are interested in league training, we can discuss further.

**Comment by zxzzz0** [comment]
@PaParaZz1 This is the exact config file mentioned in the [League Overview](https://opendilab.github.io/DI-engine/feature/league_overview_en.html?highlight=league) doc, where it said it's a demo of league. 

If it's just an agent VS bot training config, it's not a correct config for self-play with league training. Please correct this config.

We are setting up the league training. Look forward to seeing the slack channel.

**Comment by PaParaZz1** [comment]
> @PaParaZz1 This is the exact config file mentioned in the [League Overview](https://opendilab.github.io/DI-engine/feature/league_overview_en.html?highlight=league) doc, where it said it's a demo of league.
> 
> If it's just an agent VS bot training config, it's not a correct config for self-play with league training. Please correct this config.
> 
> We are setting up the league training. Look forward to seeing the slack channel.

Sorry I misunderstood something. [cpong_dqn_config](https://github.com/opendilab/DI-engine/blob/main/dizoo/competitive_rl/config/cpong_dqn_config.py) is actually the demo config for league training. Our agent VS bot config is not upload in this version. And which parts of league training is not clear enough, such as how to define own player and apply it?

**Comment by zxzzz0** [comment]
Yes. We don't understand how to apply it.
To help other people understand this and hide unnecessary details, we wrote a simple two-player game environment below for demo purpose.
```
class Game:
    def reset(self):
        return [[0, 1], [1,0]] # trivial observation
    def step(step, actions):
        if actions == [0, 0]:
            rewards = -10, -10
        elif actions == [0, 1]:
            rewards = -1, +1
        elif actions == [1, 0]:
            rewards = +1, -1
        elif actions == [1, 1]:
            rewards = 0, 0
        observations = [[0, 1], [1,0]]
        dones = True, True
        infos = None, None
        return observations, rewards, dones, infos
```
This is a simple multi agent environment for 2 players. The game will be played for just 1 step and finish. Each player has two actions in one step, `0` or `1` and they will receive rewards accordingly. Human can easily calculate the mixed strategy Nash equilibria result for this game. (select action `0` with prob `1/10` and action `1` with prob `9/10`).

After reading the League Overview doc we are still not clear which files should be changed. Code is worth a thousand words. Could you create a branch for this simple game as a demo for using opendilab's league training for custom environment? After you have the branch ready we will run the code and see if it can converge to the Nash equilibria mentioned above.

cc: @liuyuisanai for more inputs!

**Comment by zxzzz0** [comment]
We further added following 2 simple rule-based agents for evaluating the main agent during self-play.
```
class Agent1:
    def step(obs):
        return 0 # always return fixed action 0
class Agent2:
    def step(obs):
        # return actions uniformly
        return np.random.choice([0, 1], p=[0.5, 0.5]) 
```
`Agent1` always pick one fixed action. (action `0` here). `Agent2` randomly pick one action from the action set but the probability is not Nash equilibria.

If league training is going well, we expect to see the win rate between main agent against these two rule-based agents increase nearly monotonically. Please also include these two agents in your demo branch as evaluation is really an important part league training. We won't know that the main agent is indeed becoming stronger without a good set of metrics.

Feel free to let us if there are any questions.

**Comment by zxzzz0** [comment]
@liuyuisanai @PaParaZz1 Could you give comments please?

**Comment by zxzzz0** [comment]
@HansBug ping for feedbacks. Other members seemed not responsive.

**Comment by PaParaZz1** [comment]
@zxzzz0 https://join.slack.com/t/opendilab/shared_invite/zt-sqsd142v-N_l~EHLPYF1jr4c9PkqDuA You can join our slack team and discuss further

**Comment by PaParaZz1** [comment]
Anyone who interested in league training can pay attention to this branch(https://github.com/opendilab/DI-engine/tree/dev-league-demo) continually.

**Comment by PaParaZz1** [comment]
I have tested pass two groups game environments, and they both approximately converged to NE.

The first one: zero-sum

<img width="466" alt="Screen Shot 2021-07-30 at 8 51 31 PM" src="https://user-images.githubusercontent.com/33195032/127659959-cb3b8678-fc78-4514-94cf-8d4e75040731.png">
<img width="682" alt="Screen Shot 2021-07-30 at 8 51 12 PM" src="https://user-images.githubusercontent.com/33195032/127659946-f7933b39-b06f-452b-8511-c94c62a92834.png">


The second one: prisoner dilemma

<img width="472" alt="Screen Shot 2021-07-30 at 9 18 50 PM" src="https://user-images.githubusercontent.com/33195032/127660354-77274adb-766a-4ebb-9a79-635819c6d3bd.png">
<img width="675" alt="Screen Shot 2021-07-30 at 9 18 39 PM" src="https://user-images.githubusercontent.com/33195032/127660350-ff4b9405-a787-461c-813e-b01ba04e425a.png">

**Comment by PaParaZz1** [comment]
This issue has been solved in #12
